Macro use

DevinR528 opened the pull request:
---

changelog: This lint enforces Rust 2018 idiom of importing macro's directly without `#[macro_use]` fixes #5179 .

DevinR528 commented on tests/ui/macro_use_import.stderr:
```diff
@@ -0,0 +1,11 @@
+   | ^^^^^^^^^^^^
+   |
+   = note: `-D clippy::macro-use-import` implied by `-D warnings`
+   = help: remove the attribute and import the macro directly `use std::prelude::macro_name`
```
Is this an adequate auto suggestion?

llogiq replied:
No. For a suggestion we'd want to collect the macro invocations and add a `use` entry. Doing that is hard however, so we can for now do without a suggestion.

DevinR528 commented:
I can just turn the `"use {}::<macro name>"` now. I did attempt to impl the [suggestion](https://github.com/DevinR528/rust-clippy/blob/macro-use-sugg/clippy_lints/src/macro_use.rs) portion of this and got as far as collecting all the `use` imports, and macro "references" but see no way of linking them together without `ast::visit::Visitor` as done in [`librustc_resolve`](https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/check_unused.rs) they even check for the `macro_use` but only for unused imports.

JohnTitor commented on clippy_lints/src/macro_use.rs:
```diff
@@ -0,0 +1,55 @@
+                        })
+                        .collect::<String>(),
+                );
+                span_lint_and_help(ecx, MACRO_USE_IMPORT, mac_attr.span, msg, &help);
```
The current implementation doesn't consider edition, it's also triggered on 2015 edition.
Add this condition:
```rust
if cx.sess.opts.edition == Edition::Edition2018 {}
```
Also you should add `// edition:2018` flag to the test to use 2018 edition.

DevinR528 replied:
Ah yea good call thanks!

DevinR528 commented:
I ran the commands the failing tests told be to run
`tests/ui/update-references.sh '/home/devinr/aprog/rust/__forks__/rust-clippy/target/debug/test_build_base' 'regex.rs'`
and 
`tests/ui/update-references.sh '/home/devinr/aprog/rust/__forks__/rust-clippy/target/debug/test_build_base' 'single_component_path_imports.rs'`

JohnTitor commented:
A quick workaround for 4e18568f93f3f5f86ca05b4d6e6f022d3b633f9a is `cargo clean`. Also you should rebase to make CI green.

DevinR528 commented:
I `cargo clean`ed locally then `git fetch upstrem` and `git rebase upstream/master`?

DevinR528 commented:
Yay finally thanks for the help getting that working @JohnTitor !

JohnTitor commented on clippy_lints/src/macro_use.rs:
```diff
@@ -0,0 +1,58 @@
+                    span_lint_and_help(ecx, MACRO_USE_IMPORT, mac_attr.span, msg, &help);
+                }
+            }
+        }
```
This part can be indent-less by using `if_chain!` macro (see other lint implementation for usage).

DevinR528 commented:
Should I squash that commit, or the last three really into one update-reference.sh commit

DevinR528 commented:
Dang sorry I can't seem to get a good test run going they all pass locally this time there were a bunch of ICE errors. The 6 tests that failed did so with
```
error: internal compiler error: src/librustc_mir/dataflow/generic/engine.rs:315: Switch on discriminant of non-ADT

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:881:9
```

JohnTitor commented:
It isn't your fault, I can reproduce them on master. Let me take a look.

DevinR528 commented:
I have updated the new `rustc_ast` import and gotten the tests to pass I rebased when I fetched to update is it ok to `git push --force origin macro-use` to add everything up for this PR?

flip1995 commented on clippy_lints/src/macro_use.rs:
```diff
@@ -0,0 +1,59 @@
+                .iter()
+                .find(|attr| attr.ident().map(|s| s.to_string()) == Some("macro_use".to_string()));
+            then {
+                let msg = "`macro_use` attribute's are no longer needed in the Rust 2018 edition";
```
```suggestion
                let msg = "`macro_use` attributes are no longer needed in the Rust 2018 edition";
```

flip1995 commented on clippy_lints/src/macro_use.rs:
```diff
@@ -0,0 +1,59 @@
+            then {
+                let msg = "`macro_use` attribute's are no longer needed in the Rust 2018 edition";
+                    let help = format!(
+                        "remove the attribute and import the macro directly `use {}::<macro name>`",
```
Instead of this, you should use `span_lint_and_sugg` with `Applicability::HasPlaceholders`

flip1995 commented on clippy_lints/src/macro_use.rs:
```diff
@@ -0,0 +1,59 @@
+                            } else {
+                                format!("::{}", s.ident)
+                            })
+                            .collect::<String>(),
```
```suggestion
                        snippet(cx, use_tree.span, "_")
```

DevinR528 commented:
Any idea why this timed out, all the tests seemed to pass?

flip1995 commented on clippy_lints/src/macro_use.rs:
```diff
@@ -0,0 +1,53 @@
+    /// **Why is this bad?** Since the Rust 2018 edition you can import
+    /// macro's directly, this is considered idiomatic.
+    ///
+    /// **Known problems:** None.
```
Maybe add to the known problems, that the suggestion is not auto-applicable.

flip1995 commented:
Don't worry about the timeout. GHA failed here I guess. (didn't start a new run for the force push, but canceled the prev run because of a new commit)

flip1995 commented:
Thanks! Impl looks really clean

@bors r+

bors commented:
:pushpin: Commit 2840f6822340018e2d75f4a459704e7f21459122 has been approved by `flip1995`

<!-- @bors r=flip1995 2840f6822340018e2d75f4a459704e7f21459122 -->
<!-- homu: {"type":"Approved","sha":"2840f6822340018e2d75f4a459704e7f21459122","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 2840f6822340018e2d75f4a459704e7f21459122 with merge 3f8656c6cfc47080fdf53db9a6dc445b8f3b29a5...
<!-- homu: {"type":"BuildStarted","head_sha":"2840f6822340018e2d75f4a459704e7f21459122","merge_sha":"3f8656c6cfc47080fdf53db9a6dc445b8f3b29a5"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/483660268)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/483660268","builder_name":"checks-action_test"} -->

flip1995 commented:
@bors retry

bors commented:
:hourglass: Testing commit 2840f6822340018e2d75f4a459704e7f21459122 with merge b7912609fa62612d8645691a541b7f771d708ff2...
<!-- homu: {"type":"BuildStarted","head_sha":"2840f6822340018e2d75f4a459704e7f21459122","merge_sha":"b7912609fa62612d8645691a541b7f771d708ff2"} -->

bors commented:
:broken_heart: Test failed - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/484799286)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/484799286","builder_name":"checks-action_dev_test"} -->

flip1995 commented:
`cargo dev update_lints`

DevinR528 commented:
Do I need to pull first, when I ran it it made no changes

flip1995 commented:
Oh yeah, you need to rebase first probably.

DevinR528 commented:
Thats odd it fails locally but passed in CI?

flip1995 commented:
You accidentally pushed a binary file in 1601106314fec3149158a244bc85401f916d55e3. Can you remove this please?

flip1995 commented:
Can you squash some commits please? Then this is good to go.

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 597e02dcdf28e9cfbb7854bca38c28f2e4d15425 has been approved by `flip1995`

<!-- @bors r=flip1995 597e02dcdf28e9cfbb7854bca38c28f2e4d15425 -->
<!-- homu: {"type":"Approved","sha":"597e02dcdf28e9cfbb7854bca38c28f2e4d15425","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 597e02dcdf28e9cfbb7854bca38c28f2e4d15425 with merge a6a8a95cef28980c3f72de35227fb898537e23e4...
<!-- homu: {"type":"BuildStarted","head_sha":"597e02dcdf28e9cfbb7854bca38c28f2e4d15425","merge_sha":"a6a8a95cef28980c3f72de35227fb898537e23e4"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/487548160)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/487548160","builder_name":"checks-action_test"} -->

flip1995 commented:
Don't worry about the failure. This is not a issue with Clippy/this PR.

DevinR528 commented:
Thanks for all the help/input on this everyone!

flip1995 commented:
@bors retry

bors commented:
:hourglass: Testing commit 597e02dcdf28e9cfbb7854bca38c28f2e4d15425 with merge 23d2b212976d65ae72c0814d1d779d92a5d9c6fe...
<!-- homu: {"type":"BuildStarted","head_sha":"597e02dcdf28e9cfbb7854bca38c28f2e4d15425","merge_sha":"23d2b212976d65ae72c0814d1d779d92a5d9c6fe"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/488254097), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/488252919), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/488278226)
Approved by: flip1995
Pushing 23d2b212976d65ae72c0814d1d779d92a5d9c6fe to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/488278226","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/488254097","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/488252919"},"merge_sha":"23d2b212976d65ae72c0814d1d779d92a5d9c6fe"} -->
